### DISCUSSION TITLE
How can SSR be utilized in a backend-for-frontend pattern, in regards to authorization?

### DISCUSSION THREAD

**Discussion by oysandvik94** [nuxt/nuxt#17425] [Questions]
Hi. 

I'm relatively new to SSR and it's use-cases, and also not very familiar with backend-for-frontend (BFF) patterns. However, I was thinking that there's maybe a way you can utlizie SSR to do this pattern?

I guess the difficult part is maintaining a session with the clientside and the serverside. From what I understand, when you hit the clientside from you browser it hits the backend, and then back to the client, and no session is persisted? Or are there some mechanics here that I don't know about? 

It seems that it would be nice to create a serverside endpoint in Nuxt that you can use for refresh token logic for example, and store the refresh token server side, but I'm not sure how I would associate the refresh token with the user.

Also for data fetching, the normal way to do it with SSR looks to be that the data is "prefetched" on the server side before being served to the client side, which would fit well with an access token being stored on the server side, but if you need to refresh the data from the client side then that would break this pattern/flow. 

I'm really just looking for thoughts, if anyone has done something similar, or if I'm just misunderstanding several concepts and don't make sense.

**Comment by bros4president** [comment]
Hi! The very first thing is to notice the different goals of SSR and BFF. SSR is concerned primarily with display logic and a BFF is concerned primarily with business logic.

SSR might deal with some business logic, but mostly as a means to reach a display logic objective.

Also you can totally have SSR without BFF and BFF without SSR.

In the context of Nuxt 3, I'd recommend to have SSR do just that, rendering, and implement the BFF in the Nitro/server part of Nuxt (which is very independent of SSR), or another API of your choice.

Authentication falls in a bit of a gray area. You might want to do it before rendering, so the simplest way would be to implement and execute Auth logic at the beginning of SSR. On the other hand, Auth is closer to business logic, so one could argue it should be handled by the BFF. Nuxt is very flexible, and you can get the best of both worlds by implementing Auth logic in the BFF and exposing it throught its API, then calling it from a Nuxt plugin at the beginning of SSR :)

The solution I use for persisted state is cookies. Cookies are stored in the browser and shared between the backend in request/response headers. This means a first time visitor's browser would send no cookie data, so we can create a new session for him, save the session id in the cookie and then when he visits again from the same browser we can use that session id to retrieve a persisted state associated with the id.

-----------------------------------------

Disclaimer: I'm assuming here that the need for a BFF has been evaluated and accepted. If your application is responsible for just a few tasks, consumes very little external services and only has one client, what you need is a regular server application (which Nuxt handles pretty well with the Nitro engine), not a BFF.
